### ISSUE TITLE
Profile pictures

### ISSUE THREAD

**Issue by CHTJonas** [camdram/camdram#662]
I had this idea today and thought I'd open an issue for discussion. If someone has linked their user account with their public 'person' profile and their user account has a profile picture URL then that image could be displayed on their public page. This could be logically extended so that the user could customise this field with any URL from an external website (currently it is only set when they login with certain OAuth providers), and further extended to allow people to upload pictures to Camdram (although this may result in storage issues).

This would involve a pretty big change in the privacy policy as I don't think you could reasonably expect users to know their profile picture would get published. There are only 13 users who would be affected by this so it would be easy enough to contact them if we decide to move forwards with this.

```sql
SELECT * FROM acts_users WHERE person_id IS NOT NULL AND profile_picture_url IS NOT NULL;
```

Related tickets: #323 #660

**Comment by GKFX** [comment]
I'm not convinced about allowing all URLs: hotlinking has a number of downsides. Most links eventually rot and so in a few years' time there would be a number of broken images. It would also expose all users to potential tracking by the hotlinked site, which would be difficult to justify in the privacy policy; for this reason images from random URLs are currently blocked by CSP. As you say if a significant proportion of people end up displaying a profile picture that would eat into our storage significantly: there are about 15000 people on Camdram!

Database-wise, I would put the images in the people table rather than the users table if that was where we intended to make the most use of them.

**Comment by philosophicles** [comment]
I quite like this idea, provided we handle the privacy policy changes / notifications properly; and to be honest we should probably make it a specific opt-in feature anyway, not on-by-default if there's a person<-->user link established. 

I think we should only support profile pics served by one of the OAuth providers, not random URLs or uploads direct to Camdram. That prevents a lot of potential abuse angles. We could _maybe_ add support for [Gravatar](https://en.gravatar.com/) I guess, as one extra image source. 

At the moment, if somebody uses multiple OAuth providers, I think we don't offer any way for them to choose which provider's profile pic is used for their Camdram user account. If adding this Person usage, we should probably offer that choice - e.g. someone has a very informal Facebook profile pic, and a professional headshot on their Twitter that they'd want to use.

**Comment by philosophicles** [comment]
Storage wise, so long as we can spare a bit of disk space, I wouldn't worry too much. We should appropriately downscale images before storing, if applicable. We might have 15k people, but not quite 4800 users, so that's the lower bound that matters more. Even if all of them had a 20 kB image in our database, that's under 100MB.

**Comment by CHTJonas** [comment]
Okay let's stick to known sites then; that way we can avoid uploads and manage CSP. At a glance it looks like Facebook and (the now defunct) Google+ are the only OAuth providers that are storing profile picture URLs in the database. Gravatar should be easy enough to support but worth checking to see if Twitter has a similar picture feature.

**Comment by philosophicles** [comment]
Yes, Twitter's API also allows this it seems: https://developer.twitter.com/en/docs/accounts-and-users/user-profile-images-and-banners. 

The specific endpoint required being `GET users/show`: https://developer.twitter.com/en/docs/accounts-and-users/follow-search-get-users/api-reference/get-users-show.

**Comment by philosophicles** [comment]
☝️ have raised a new ticket for implementing retrieval of profile picture (URLs) from Twitter, since that's a rather different (and less controversial?) thing to this suggested feature.
